core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -416,13 +416,16 @@ private[zio] final class FiberContext[E, A](
     }
 
   @inline
-  private[this] final def shouldDie: Boolean = noInterrupt == 0 && state.get.interrupted
+  private[this] final def interruptible: Boolean =
+    interruptStatus.peekOrElse(true)
 
   @inline
-  private[this] final def allowRecovery: Boolean = {
-    val currentState = state.get
-    !currentState.interrupted || !currentState.terminating && noInterrupt != 0
-  }
+  private[this] final def shouldInterrupt: Boolean =
+    state.get.interrupted && interruptible && !state.get.terminating
+
+  @inline
+  private[this] final def allowRecovery: Boolean =
+    !(state.get.interrupted && interruptible)
```

Comment:
Isn't there a case where recovery isn't permitted during termination, e.g. when `terminate(IO.interrupt)` is invoked?

Reply:
We have to think this through, but no tests broke, partially because I changed `shouldDie` (`shouldInterrupt`) to always return `false` if in the process of terminating, and I only turned on `terminating` in the sole case of termination from the evaluation loop (kill no longer does that, rather, it instructs the main loop to restart with interruption, so when it interrupts, it will then go into the termination state).

I think it's correct and consistent that inside any uninterruptible section, you can catch everything you can outside, PLUS interruption. After all, you can already "catch" defects if you want with `run` / `sandbox` / `foldCauseM`. So there's no change between uninterruptible and interruptible there. These semantics seem clear, clean, and consistent, but of course more time spent thinking about it is always a good thing.